xdsl/transforms/canonicalization_patterns/riscv.py

```diff
@@ -70,6 +71,49 @@ def match_and_rewrite(self, op: riscv.MulOp, rewriter: PatternRewriter) -> None:
             rewriter.replace_matched_op(riscv.MVOp(op.rs2, rd=rd))
 
 
+class MulImmediatePower2(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: riscv.MulOp, rewriter: PatternRewriter) -> None:
+        lhs: int | None = None
+        rhs: int | None = None
+
+        if (rs1 := get_constant_value(op.rs1)) is not None:
+            lhs = rs1.value.data
+        if (rs2 := get_constant_value(op.rs2)) is not None:
+            rhs = rs2.value.data
+
+        rd = cast(riscv.IntRegisterType, op.rd.type)
+
+        if rhs is None:
+            return
+
+        # get shl immediate
+        i = int(log2(abs(rhs)))
+
+        match (lhs, rhs < 0):
+            case None, False:
+                rewriter.replace_matched_op(
+                    riscv.SlliOp(op.rs1, i, rd=rd, comment=op.comment)
+                )
+            case int(), False:
+                rewriter.replace_matched_op(
+                    riscv.LiOp(lhs << i, rd=rd, comment=op.comment)
+                )
```

Comment:
won't this already be done by the pattern that multiplies constant values together? I don't think we need to handle this case here